### PR TITLE
Ignore some errors on leaving rooms, add new error enum. Fixes #307

### LINKED ISSUE #307: Forget rooms even when leaving them fails
### Description

It's possible to get into state where you have a room that you cannot leave / forget.

### Steps to reproduce

No reliable way found yet, but try:

- Send IRC private message to user via bridge
- Accept it
- ???
- Leave room
- Observe leaving not possible (Using Spectral client, should not be client dependent)

There are also issues with leaving rooms in Riot, so probably a Synapse bug is related.

### Implementation

According to Kitsune:

So the proper fix (or workaround, if it's a server-side problem) is:
1. Introduce a separate enum code for M_UNKNOWN, so that you could separate this error from generic IncorrectRequestErrors. Alternatively, check for M_UNKNOWN in JSON in leaveJob->errorString().
2. Factor out the code that flushes the room from the roomMap to a separate method inside Connection::Private
3. Check the status for the code from (1), and call the method from (2) if the condition holds.

(I'm already working on this so feel free to assign it to me)

### PR DESCRIPTION
**Pull request by vranki**
Tested with Spectral as client and the problematic rooms are forgot properly after this fix.

### CODE REVIEW

File: lib/connection.cpp [hunk 1]
```diff
@@ -737,29 +738,36 @@ ForgetRoomJob* Connection::forgetRoom(const QString& id)
...
         auto leaveJob = room->leaveRoom();
-        connect(leaveJob, &BaseJob::success, this, [this, forgetJob, room] {
-            forgetJob->start(connectionData());
-            // If the matching /sync response hasn't arrived yet, mark the room
-            // for explicit deletion
-            if (room->joinState() != JoinState::Leave)
-                d->roomIdsToForget.push_back(room->id());
+        connect(leaveJob, &BaseJob::result, this, [this, leaveJob, forgetJob, room] {
+            // After leave, continue if there is no error or the room id is not found (IncorrectRequestError)
+            if(!leaveJob->error() || leaveJob->error() == BaseJob::StatusCode::IncorrectRequestError || leaveJob->error() == BaseJob::StatusCode::UnknownError) {
```
Comment (KitsuneRal) [inline lib/connection.cpp]: Please don't ignore `IncorrectRequestError` - you can't imagine how many times this code helped me debug the code already.
```suggestion
            if(!leaveJob->error() ||
                    leaveJob->error() == BaseJob::StatusCode::UnknownError) {
```

File: lib/connection.cpp [hunk 2]
```diff
@@ -737,29 +738,36 @@ ForgetRoomJob* Connection::forgetRoom(const QString& id)
...
-            if (auto r = d->roomMap.take({ id, f }))
-            {
-                qCDebug(MAIN) << "Room" << r->objectName()
-                              << "in state" << toCString(r->joinState())
-                              << "will be deleted";
-                emit r->beforeDestruction(r);
-                r->deleteLater();
-            }
+        // Leave room in case of success, or room not known by server
+        if(!forgetJob->error() || forgetJob->error() == BaseJob::StatusCode::IncorrectRequestError || forgetJob->error() == BaseJob::StatusCode::UnknownError) {
```
Comment (KitsuneRal) [inline lib/connection.cpp]: I don't think this job ever returns `M_UNKNOWN` on unknown rooms - it returns 200 instead, no?

File: lib/jobs/basejob.h [hunk 3]
```diff
@@ -67,6 +67,7 @@ namespace QMatrixClient
                 , UnsupportedRoomVersionError
                 , NetworkAuthRequiredError
                 , UserConsentRequiredError
+                , UnknownError // Unknown room or other item (M_UNKNOWN)
```
Comment (KitsuneRal) [inline lib/jobs/basejob.h]: Can we please make it `UnknownObjectError` or something? `UnknownError` looks really ambiguous.

File: lib/connection.cpp (hunk 1 above)
Comment (vranki) [reply lib/connection.cpp]: Removed IncorrectRequestError in next PR

**Comment by vranki** [comment]
Work on this continued in PR https://github.com/quotient-im/libQuotient/pull/327/

**Pull request closed without merging**